test/cpp/api/parallel.cpp

```diff
@@ -0,0 +1,179 @@
+#include <catch.hpp>
+
+#include <torch/csrc/autograd/functions/comm.h>
+#include <torch/nn/module.h>
+#include <torch/nn/modules/linear.h>
+#include <torch/nn/parallel/data_parallel.h>
+#include <torch/nn/pimpl.h>
+#include <torch/tensor.h>
+
+#include <iostream>
+#include <memory>
+#include <utility>
+#include <vector>
+
+using namespace torch::autograd;
+using namespace torch::nn;
+
+TEST_CASE("Parallel/DifferentiableScatter", "[cuda]") {
+  Scatter scatter(
+      {torch::Device(torch::kCUDA, 0), torch::Device(torch::kCUDA, 1)});
+
+  auto input = torch::ones(10, torch::requires_grad(true));
+  auto output = scatter.apply({input});
+
+  REQUIRE(output.size() == 2);
+  REQUIRE(output[0].size(0) == 5);
+  REQUIRE(output[1].size(0) == 5);
+
+  auto sum = output[0].to({torch::kCUDA, 1}) + output[1];
+  sum.backward();
+
+  REQUIRE(input.grad().defined());
+  REQUIRE(input.grad().device().is_cpu());
+  REQUIRE(input.grad().sum().toCInt() == 10);
+}
+
+TEST_CASE("Parallel/DifferentiableGather", "[cuda]") {
+  Gather gather(torch::Device(torch::kCUDA, 1));
+
+  auto a = torch::ones(5, torch::requires_grad(true).device({torch::kCUDA, 0}));
+  auto b = torch::ones(5, torch::requires_grad(true).device({torch::kCUDA, 1}));
+  auto output = gather.apply({a, b});
+
+  REQUIRE(output.size() == 1);
+  REQUIRE(output[0].size(0) == 10);
+  REQUIRE(output[0].device() == torch::Device(torch::kCUDA, 1));
+
+  output[0].backward();
+
+  REQUIRE(a.grad().defined());
+  REQUIRE(a.grad().device() == torch::Device(torch::kCUDA, 0));
+  REQUIRE(a.grad().sum().toCInt() == 5);
+
+  REQUIRE(b.grad().defined());
+  REQUIRE(b.grad().device() == torch::Device(torch::kCUDA, 1));
+  REQUIRE(b.grad().sum().toCInt() == 5);
```

Comment:
We should have sth like `.eq(1).all()` in ATen, and it's a stronger check than this. Also, ideally you'd pass in a random output and check that the grads here are just slices